### PR TITLE
Js widgets

### PR DESCRIPTION
**Pull request by SeijiEmery**
Wraps 2d overlays with a retained-mode widget library, which provides event-driven gui programming and basic layouts and controls.

### CODE REVIEW

**Comment by SeijiEmery** [comment]
Quick screenshot w/ debugging UI active:
screen shot 2015-08-12 at 4 04 13 pm

File: examples/example/widgets-example.js [hunk 1]
```diff
@@ -0,0 +1,442 @@
...
+var panelY = 500;
+var panelWidth = 50;
+var panelHeight = 210;
+
+// var mainPanel = new UIPanel(panelX, panelY, panelWidth, panelHeight);
+// var systemViewButton = mainPanel.addIcon('solarsystems');
+// var zoomButton = mainPanel.addIcon('magnifier');
+// var satelliteButton = mainPanel.addIcon('satellite');
+// var settingsButton = mainPanel.addIcon('settings');
+// var stopButton = mainPanel.addIcon('close');
```
Comment (huffman) [inline examples/example/widgets-example.js]: This looks like old code (not just commented out examples), let's remove these if so

File: examples/example/widgets-example.js [hunk 2]
```diff
@@ -0,0 +1,442 @@
+// 
+//  widgets-example.js
+//  games
```
Comment (huffman) [inline examples/example/widgets-example.js]: Remove this? Also add your name :D

File: examples/example/widgets-example.js [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+}
+
+function makeDraggable (panel, target) {
+    if (!target)
+        target = panel;
+
+    var dragStart = null;
+    var initialPos = null;
+
+    panel.addAction('onDragBegin', function (event) {
```
Comment (huffman) [inline examples/example/widgets-example.js]: I think the name `addAction` is a little confusing - we aren't adding a new event -  `addActionListener` might be clearer

File: examples/example/widgets-example.js [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
+        UI.handleMousePress(event);
+    },
+    onMouseRelease: function (event) {
+        clearDrag(event);
+        UI.handleMouseRelease(event);
+    }
+};
+Controller.mousePressEvent.connect(inputHandler.onMousePress);
+Controller.mouseMoveEvent.connect(inputHandler.onMouseMove);
+Controller.mouseReleaseEvent.connect(inputHandler.onMouseRelease);
```
Comment (huffman) [inline examples/example/widgets-example.js]: Any reason not to have UI register these for itself?

File: examples/example/widgets-example.js [hunk 5]
```diff
@@ -0,0 +1,442 @@
...
+        dragged = dragAction;
+    }
+    this.updateDrag = function (event) {
+        if (dragged) {
+            print("Update drag");
+            dragged.updateDrag(event);
+        }
+    }
+    this.clearDrag = function (event) {
+        if (dragged)
```
Comment (huffman) [inline examples/example/widgets-example.js]: style nitpick - the guide specified that all blocks should be wrapped in curly braces even if they aren't required.

File: examples/libraries/uiwidgets.js [hunk 6]
```diff
@@ -0,0 +1,1026 @@
...
+//	uiwidgets.js
+//	examples/libraries
+//
+//	Created by Seiji Emery, 8/10/15
+//	Copyright 2015 High Fidelity, Inc
+//
+//  Distributed under the Apache License, Version 2.0.
+//  See the accompanying file LICENSE or http://www.apache.org/licenses/LICENSE-2.0.html
+//
+
```
Comment (huffman) [inline examples/libraries/uiwidgets.js]: Replace tabs with 4 spaces

File: examples/libraries/uiwidgets.js [hunk 7]
```diff
@@ -0,0 +1,1026 @@
...
+		throw new Error("assertion failed: " + expr + " (" + cond + ")" + "\n" +
+			"Called from: " + callstack + " " +
+			"Traceback: \n\t" + __trace.join("\n\t"));
+	}
+}
+var traceEnter = function(fcn) {
+	var l = __trace.length;
+	// print("TRACE ENTER: " + (l+1));
+	s = "";
+	for (var i = 0; i < __traceDepth+1; ++i)
```
Comment (huffman) [inline examples/libraries/uiwidgets.js]: braces

**Comment by huffman** [comment]
Nice work! I know you need to get this in, so I'll merge it because it's just JS files. Let's get some of those other changes in in a future PR.

**Comment by PhilipRosedale** [comment]
Are there docs for this?  Where can I read what new capabilities this introduces and how they work?

**Pull request merged**